src/coreclr/jit/emitriscv64.cpp

```diff
@@ -2430,77 +2445,152 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
                 {
                     case 8:
                     {
-                        regNumber reg2 = id->idReg2();
+                        assert((INS_blt <= ins && ins <= INS_bgeu) || (INS_beq == ins) || (INS_bne == ins) ||
+                               (INS_bnez == ins) || (INS_beqz == ins));
+                        assert(isValidSimm21(imm));
+                        assert((emitInsCode(INS_bne) & 0xefff) == emitInsCode(INS_beq));
+                        assert((emitInsCode(INS_bge) & 0xefff) == emitInsCode(INS_blt));
+                        assert((emitInsCode(INS_bgeu) & 0xefff) == emitInsCode(INS_bltu));
+
+                        regNumber reg2 = REG_R0;
+                        if (INS_beqz != ins && INS_bnez != ins)
+                            reg2 = id->idReg2();
+                        code     = emitInsCode(ins) ^ 0x1000;
+                        code |= (code_t)reg1 << 15; /* rj */
+                        code |= (code_t)reg2 << 20; /* rd */
+                        code |= 0x8 << 7;
+                        *(code_t*)dstRW = code;
+                        dstRW += 4;
+
+                        code = emitInsCode(INS_jal);
+                        code |= ((imm >> 12) & 0xff) << 12;
+                        code |= ((imm >> 11) & 0x1) << 20;
+                        code |= ((imm >> 1) & 0x3ff) << 21;
+                        code |= ((imm >> 20) & 0x1) << 31;
+
+                        *(code_t*)dstRW = code;
+                        dstRW += 4;
+                        break;
+                    }
+                    case 24:
+                    {
+                        assert(ins == INS_j || ins == INS_jal);
+                        imm               = imm - 2 * 4;
+                        regNumber tmpReg1 = REG_RA;
+                        ssize_t   high    = ((imm + 0x80000000) >> 32) & 0xffffffff;
+                        code              = emitInsCode(INS_lui);
+                        code |= (code_t)tmpReg1 << 7;
+                        code |= ((code_t)((high + 0x800) >> 12) & 0xfffff) << 12;
+
+                        *(code_t*)dstRW = code;
+                        dstRW += 4;
+
+                        code = emitInsCode(INS_addi);
+                        code |= (code_t)tmpReg1 << 7;
+                        code |= (code_t)tmpReg1 << 15;
+                        code |= (code_t)(high & 0xfff) << 20;
+                        *(code_t*)dstRW = code;
+                        dstRW += 4;
+
+                        code = emitInsCode(INS_slli);
+                        code |= (code_t)tmpReg1 << 7;
+                        code |= (code_t)tmpReg1 << 15;
+                        code |= (code_t)32 << 20;
+                        *(code_t*)dstRW = code;
+                        dstRW += 4;
+
+                        regNumber tmpReg2 = codeGen->rsGetRsvdReg();
+                        ssize_t   low     = imm & 0xffffffff;
+                        code              = emitInsCode(INS_auipc);
```

Comment:
Are you sure that `auipc` should be there? It build PC relative address, which could be more then 32 bit and may corrupt `tmpReg1` after addition with it.

Reply:
In those cases, it converts JAL instruction to JALR instruction because the offset in JAL is too huge to include in JAL encoding.
```
jal: x[rd] = pc+4; pc += sext(offset)
jalr: t =pc+4; pc=(x[rs1]+sext(offset))&∼1; x[rd]=t
```
In JAL, offset is the difference between current address and target address. However, in JALR, the value in rs1 should be target address. To convert JAL to JALR, the offset should be converted to target address in rs1. So I set PC + offset to tmpReg2 (except offset[11:0]. It is set in jalr instruction to reduce instruction count.).
If I am wrong, please let me know.
Thank you.

Reply:
(Will delay merging this PR to be sure this is ok.)

Reply:
@jakobbotsch I found another issues. After I fix them, then I will request your review again. Thank you.

Reply:
My complain was about `auipc`, but I checked and it doesn't use sign extension and always fit to 32 bits.

Also could add comment that `imm` there is relative offset.

Reply:
@jakobbotsch
I discussed the issue with @alpencolt and he said he doesn't see issues now. Thank you @alpencolt
I fixed the dest reg in jalr and checked the generated codes on `./JIT/jit64/opt/cse/HugeField2/HugeField2.sh` test (It is a single test that converts jal to jalr).
Could you review again?
Thanks a lot!